### DISCUSSION TITLE
How is spdlog so fast ??

### DISCUSSION THREAD

**Discussion by beemibrahim** [gabime/spdlog#2370] [Q&A]
I have been trying to make my own logging framework , and I realized that my single threaded api is really slow , how is spdlog so fast (single - threaded).

**Comment by tt4g** [comment]
Without measuring benchmarks, it is difficult to determine the cause of slow single-thread API, but in general I/O operations tend to be slower.

One of spdlog's tricks is to delay I/O flushes. This trick reduces I/O latency and thus improves code performance.

**Reply by beemibrahim** [reply]
Improves code performance by how much ??

**Reply by tt4g** [reply]
Without measuring benchmarks, it is impossible to know how much performance will improve.

I don't know where the bottleneck is for libraries other than spdlog.
This is because it is impossible to know where the cause of the slow execution speed lies without a thorough investigation.

If the problem is with your own library, you should open an issue in your own repository or ask a question on StackOverflow, not in the spdlog repository.

**Comment by beemibrahim** [comment]
Can Please tell you more tricks

**Comment by beemibrahim** [comment]
There should be more tricks

**Comment by USAFrenzy** [comment]
There's really no better trick than just measuring and profiling your code - one tip is to mess with/add functionality to increase the file buffer used; while spdlog delays the OS call for flushing, file buffers are still flushed to disk when they are full. Therefore, increasing the size of that buffer for your use case will also speed times up quite a bit. The default size is generally usually 8k so you could use that as a starting point for using setvbuf() for instance. Spdlog's speed in formatting the timestamp portion heavily relies on manual structures hand-rolled to mimic what you would get from a strftime call; there's really nothing faster than a specific hand-rolled version for your use case rather than relying on generic broad functions as a general rule-of-thumb. The more abstracted your function calls are, the more likely there's some overhead that you incur; so the lower you get to OS-specific code you can get, the faster your code will tend to be (not always, but usually).
